File: day13/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

type (
	Pos struct {
		x int
		y int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	points := map[Pos]struct{}{}
	trackPoints := true
	first := true
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			trackPoints = false
			continue
		}
		if trackPoints {
			arr := strings.SplitN(line, ",", 2)
			if len(arr) != 2 {
				log.Fatalln("Invalid line")
			}
			x, err := strconv.Atoi(arr[0])
			if err != nil {
				log.Fatal(err)
			}
			y, err := strconv.Atoi(arr[1])
			if err != nil {
				log.Fatal(err)
			}
			points[Pos{
				x: x,
				y: y,
			}] = struct{}{}
			continue
		}
		words := strings.Fields(line)
		if len(words) != 3 {
			log.Fatalln("Invalid line")
		}
		arr := strings.SplitN(words[2], "=", 2)
		if len(arr) != 2 {
			log.Fatalln("Invalid line")
		}
		axis := arr[0]
		axisval, err := strconv.Atoi(arr[1])
		if err != nil {
			log.Fatal(err)
		}
		points = fold(points, axis == "y", axisval)
		if first {
			first = false
			fmt.Println(len(points))
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	maxx, maxy := findMax(points)
	grid := make([][]byte, maxy)
	for i := range grid {
		row := make([]byte, maxx)
		for j := range row {
			row[j] = ' '
		}
		grid[i] = row
	}
	for k := range points {
		grid[k.y][k.x] = '#'
	}

	for _, i := range grid {
		fmt.Println(string(i))
	}
}

func fold(points map[Pos]struct{}, yaxis bool, axisval int) map[Pos]struct{} {
	next := map[Pos]struct{}{}
	for k := range points {
		if yaxis {
			if k.y <= axisval {
				next[k] = struct{}{}
				continue
			}
			next[Pos{
				x: k.x,
				y: axisval - (k.y - axisval),
			}] = struct{}{}
		} else {
			if k.x <= axisval {
				next[k] = struct{}{}
				continue
			}
			next[Pos{
				x: axisval - (k.x - axisval),
				y: k.y,
			}] = struct{}{}
		}
	}
	return next
}

func findMax(points map[Pos]struct{}) (int, int) {
	x := 0
	y := 0
	for k := range points {
		if k.x > x {
			x = k.x
		}
		if k.y > y {
			y = k.y
		}
	}
	return x + 1, y + 1
}

File: day13/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

#[derive(PartialEq, Eq, Hash, Clone, Copy)]
struct Pos(usize, usize);

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut points = HashSet::new();
    let mut track_points = true;
    let mut first = true;
    for line in reader.lines() {
        let line = line?;
        if line == "" {
            track_points = false;
            continue;
        }
        if track_points {
            if let Some((xs, ys)) = line.split_once(",") {
                points.insert(Pos(xs.parse::<usize>()?, ys.parse::<usize>()?));
            } else {
                return Err("Invalid line".into());
            }
            continue;
        }
        let eqn = match line.rsplit_once(" ") {
            Some((_, eqn)) => eqn,
            _ => return Err("Invalid line".into()),
        };
        let (yaxis, val) = match eqn.split_once("=") {
            Some((axis, val)) => (axis == "y", val.parse::<usize>()?),
            _ => return Err("Invalid line".into()),
        };
        points = fold(points, yaxis, val);
        if first {
            first = false;
            println!("Part 1: {}", points.len());
        }
    }
    let (mx, my) = find_max(&points);
    let mut grid = (0..=my)
        .into_iter()
        .map(|_| vec![b' '; mx + 1])
        .collect::<Vec<_>>();
    for Pos(x, y) in points {
        grid[y][x] = b'#';
    }

    println!("Part 2:");
    for i in grid {
        println!("{}", String::from_utf8(i)?);
    }
    Ok(())
}

fn fold(points: HashSet<Pos>, yaxis: bool, axis: usize) -> HashSet<Pos> {
    let mut next = HashSet::new();
    for Pos(x, y) in points {
        if yaxis {
            if y <= axis {
                next.insert(Pos(x, y));
                continue;
            }
            next.insert(Pos(x, axis - (y - axis)));
        } else {
            if x <= axis {
                next.insert(Pos(x, y));
                continue;
            }
            next.insert(Pos(axis - (x - axis), y));
        }
    }
    next
}

fn find_max(points: &HashSet<Pos>) -> (usize, usize) {
    points.iter().fold((0, 0), |(mx, my), &Pos(x, y)| {
        (if x > mx { x } else { mx }, if y > my { y } else { my })
    })
}
